Tools/Scripts/layout-test-failure-analysis

```diff
@@ -0,0 +1,291 @@
+    info("Parsing into chronological list of tests...")
+    data = raw.split("\n")
+    tests = []
+    i = 0
+    for line in data:
+        m = re.findall(r"worker\/(\d+) ([^\"\s].*\.(htm|html|pl|py|svg|xht|xhtml|xml)([?]?.*)) (passed|failed)", line)
+        if len(m) > 0:
+            # continue
+            i += 1
+            test_name = m[0][1]
+            tests.append(test_name)
+            if not long and test_name == target_test:
+                info(f'Found requested test (last of {i}), skipping remaining tests.')
+                break
+    print(f'Processed {i} tests total.')
+    return "\n".join(tests)
+
+
+# Parses a raw stdio log into a dict of lists containing test lists.
+# Used when exporting one worker, or all as separate files.
+def parse_raw_stdio_into_workers(raw, target_test='', long=False):
+    info("Parsing into workers...")
+    data = raw.split("\n")
+    workers = {}
+    target_worker = -1
+    i = 0
+    for line in data:
+        m = re.findall(r"worker\/(\d+) ([^\"\s]+.+.(htm|html|pl|py|svg|xht|xhtml|xml)([?]?.*)) (passed|failed)", line)
+        if len(m) > 0:
+            worker, test_name = (str(m[0][0]), m[0][1])
+            i += 1
+            if type(workers.get(worker)) != list:
+                workers.update({f'{worker}': []})
+            workers[worker].append(test_name)
+            if test_name == target_test:
+                info(f'Found requested test in worker {worker} (test run #{len(workers[worker])} in that worker).')
+                target_worker = worker
+                if long:
+                    info(f'Continuing to sort tests by worker (--long).')
+                    continue
+                break
+    print(f'Finished parsing tests ({i} total).')
+    return (workers, int(target_worker))
+
+
+# Loads the stdio input for analysis from a file.
+# Returns a str with the contents of the stdio log.
+def load_from_file(path):
+    info("Loading data from file...")
+    if not os.path.isfile(path):
+        error(f'Could not find a file at the specified path "{path}".')
+        graceful_exit(os.EX_DATAERR)
+    text = ""
+    with open(path) as file:
+        for line in file:
+            text += line
+    return text
+
+
+# Loads the stdio input for analysis from a build URL.
+# Returns a str with the contents of the stdio log.
+def load_from_url(url):
+    url = url.replace('/#/', '/api/v2/') + ("/raw" if "/logs/stdio" in url else "/steps/layout-test/logs/stdio/raw")
+    try: # validation
+        uobj = urlparse(url)
+        if not all([uobj.scheme, uobj.netloc]):
+            raise IOError()
+    except:
+        error("Invalid URL specified: %s.", url)
+        graceful_exit(os.EX_DATAERR)
+
+    info("Loading from URL (this may take a moment)...")
+
+    try: # more validation
+        data = requests.get(url)
+    except Exception as e:
+        error('Unable to load data: ' + e.args[0] + '. Check your spelling and try again.')
+        graceful_exit(os.EX_DATAERR)
```

Comment:
This is a good example of the right way to catch a general exception.

Worth noting that `sys.exit` itself actually raises an exception. This code isn't broken by that, but it's a bit of a surprising fact that can bite defensive error handling like this which attempts to manage general exceptions.

Reply:
Thanks for the info on that! Does `sys.exit` raising an exception break this, and/or does it require any changes?

Reply:
`sys.exit` raising an exception won't break this, just wanted to call that fact out since I can see refactors which would be bit by that.

Reply:
Gotcha, I'll add a comment to that effect. Thanks for the tip!